src/main/java/sparksoniq/semantics/visitor/RuntimeIteratorVisitor.java

```diff
@@ -357,27 +362,39 @@ public RuntimeIterator visitFunctionDeclaration(FunctionDeclaration expression,
     public RuntimeIterator visitFunctionCall(FunctionCall expression, RuntimeIterator argument) {
         List<RuntimeIterator> arguments = new ArrayList<>();
         IteratorMetadata iteratorMetadata = createIteratorMetadata(expression);
+        boolean isPartialApplication = false;
         for (Expression arg : expression.getArguments()) {
-            arguments.add(this.visit(arg, argument));
+            if (arg instanceof ArgumentPlaceholder) {
+                arguments.add(null);
+                isPartialApplication = true;
+            } else {
+                arguments.add(this.visit(arg, argument));
+            }
         }
         String fnName = expression.getFunctionName();
         int arity = arguments.size();
         FunctionIdentifier fnIdentifier = new FunctionIdentifier(fnName, arity);
 
         try {
             Class<? extends RuntimeIterator> functionClass = Functions.getBuiltInFunction(fnIdentifier, createIteratorMetadata(expression));
-            Constructor<? extends RuntimeIterator> ctor = functionClass.getConstructor(List.class, IteratorMetadata.class);
-            return ctor.newInstance(arguments, iteratorMetadata);
-        } catch (Exception ex1) {
-            if (ex1 instanceof UnknownFunctionCallException) {
-                return new UserDefinedFunctionCallIterator(
-                        fnIdentifier,
-                        arguments,
-                        iteratorMetadata
+            if (isPartialApplication) {
+                throw new UnsupportedFeatureException(
+                        "Partial application on built-in functions are not supported.",
+                        expression.getMetadata()
                 );
-            } else {
-                throw new RuntimeException(ex1.getMessage());
             }
+            Constructor<? extends RuntimeIterator> ctor = functionClass.getConstructor(List.class, IteratorMetadata.class);
+            return ctor.newInstance(arguments, iteratorMetadata);
+        } catch (UnknownFunctionCallException e) {
```

Comment:
Can you make a note/issue for the future to use a test rather than use exceptions to switch between builtin/UDF? It makes a difference in performance. Exceptions should only be used for actual errors.

It's good for now (correctness and design of test cases had priority) but we should keep it in mind.

We could talk about this next week to clean up the mechanism.